PEP 756: PyUnicode_Export() (#3960)
python/peps: peps/pep-0756.rst

PEP: 756
Title: Add PyUnicode_Export() and PyUnicode_Import() C functions
Author: Victor Stinner <vstinner@python.org>
PEP-Delegate: C API Working Group
Status: Draft
Type: Standards Track
Created: 13-Sep-2024
Python-Version: 3.14

.. highlight:: c

Abstract
========

Add functions to the limited C API version 3.14:

* ``PyUnicode_Export()``: export a Python str object as a ``Py_buffer``
  view.
* ``PyUnicode_Import()``: import a Python str object.

> In general, ``PyUnicode_Export()`` has a O(1) complexity: no memory

Reviewer: ```suggestion
In general, ``PyUnicode_Export()`` has an *O*\ (1) complexity: no memory
```
https://devguide.python.org/documentation/markup/#typographic-conventions

copy is needed. See the :ref:`specification <export-complexity>` for
cases when a copy is needed.

Rationale
=========

PEP 393
-------

:pep:`393` "Flexible String Representation" changed string internals in
Python 3.3 to use three formats:

> * UCS1: Unicode range [U+0000; U+00ff], 1 byte per character
> * UCS2: Unicode range [U+0000; U+ffff], 2 bytes per character
> * UCS4: Unicode range [U+0000; U+10ffff], 4 bytes per character

Reviewer: Like movies, I think the first one was unnumbered? The others are hyphented:

https://en.wikipedia.org/wiki/Universal_Coded_Character_Set

```suggestion
* UCS: Unicode range [U+0000; U+00ff], 1 byte per character
* UCS-2: Unicode range [U+0000; U+ffff], 2 bytes per character
* UCS-4: Unicode range [U+0000; U+10ffff], 4 bytes per character
```

A Python ``str`` object must always use the most compact format. For

> example, a string which only contain ASCII characters must use the UCS1
> format.

Reviewer: ```suggestion
example, a string which only contains ASCII characters must use the original
UCS format.
```

The ``PyUnicode_KIND()`` function can be used to know the format used by
a string:

> * ``PyUnicode_1BYTE_KIND``: UCS1, 1 byte per character
> * ``PyUnicode_2BYTE_KIND``: UCS2, 2 bytes per character
> * ``PyUnicode_4BYTE_KIND``: UCS4, 4 bytes per character

Reviewer: ```suggestion
* ``PyUnicode_1BYTE_KIND``: UCS, 1 byte per character
* ``PyUnicode_2BYTE_KIND``: UCS-2, 2 bytes per character
* ``PyUnicode_4BYTE_KIND``: UCS-4, 4 bytes per character
```

> Then, one of the following function can be used to access data:

Reviewer: ```suggestion
Then, one of the following functions can be used to access data:
```

> * UCS1: ``PyUnicode_1BYTE_DATA()``
> * UCS2: ``PyUnicode_2BYTE_DATA()``
> * UCS4: ``PyUnicode_4BYTE_DATA()``

Reviewer: ```suggestion
* UCS: ``PyUnicode_1BYTE_DATA()``
* UCS-2: ``PyUnicode_2BYTE_DATA()``
* UCS-4: ``PyUnicode_4BYTE_DATA()``
```

> To get best performance, a C extension should have 3 code paths for each

Reviewer: ```suggestion
To get the best performance, a C extension should have 3 code paths for each
```

of these 3 string native formats.

Limited C API
-------------

> PEP 393 functions such as ``PyUnicode_KIND()`` and

Reviewer: ```suggestion
:pep:`393` functions such as ``PyUnicode_KIND()`` and
```

``PyUnicode_1BYTE_DATA()`` are excluded from the limited C API. It's not
possible to write code specialized for UCS formats. A C extension using
the limited C API can only use less efficient code paths and string
formats.

> For example, the Markupsafe project has a C extension specialized for

Reviewer: ```suggestion
For example, the MarkupSafe project has a C extension specialized for
```

        Py_ssize_t nbytes,
        int32_t format);

    #define PyUnicode_FORMAT_UCS1  0x01   // Py_UCS1*
    #define PyUnicode_FORMAT_UCS2  0x02   // Py_UCS2*
    #define PyUnicode_FORMAT_UCS4  0x04   // Py_UCS4*
    #define PyUnicode_FORMAT_UTF8  0x08   // char*
    #define PyUnicode_FORMAT_ASCII 0x10   // char* (ASCII string)

PyUnicode_Export()
------------------

API: ``int32_t PyUnicode_Export(PyObject *unicode, int32_t requested_formats, Py_buffer *view)``.

Export the contents of the *unicode* string in one of the *requested_formats*.

* On success, fill *view*, and return a format (greater than ``0``).
* On error, set an exception, and return ``-1``.
  *view* is left unchanged.

After a successful call to ``PyUnicode_Export()``,
the *view* buffer must be released by ``PyBuffer_Release()``.
The contents of the buffer are valid until they are released.

The buffer is read-only and must not be modified.

*unicode* and *view* must not be NULL.

Available formats:

===================================  ========  ===========================
Constant Identifier                  Value     Description
===================================  ========  ===========================

> ``PyUnicode_FORMAT_UCS1``            ``0x01``  UCS-1 string (``Py_UCS1*``)

Reviewer: ```suggestion
``PyUnicode_FORMAT_UCS1``            ``0x01``  UCS string (``Py_UCS1*``)
```

``PyUnicode_FORMAT_UCS2``            ``0x02``  UCS-2 string (``Py_UCS2*``)
``PyUnicode_FORMAT_UCS4``            ``0x04``  UCS-4 string (``Py_UCS4*``)
``PyUnicode_FORMAT_UTF8``            ``0x08``  UTF-8 string (``char*``)
``PyUnicode_FORMAT_ASCII``           ``0x10``  ASCII string (``Py_UCS1*``)
===================================  ========  ===========================

UCS-2 and UCS-4 use the native byte order.

*requested_formats* can be a single format or a bitwise combination of the
formats in the table above.
On success, the returned format will be set to a single one of the requested
flags.

Note that future versions of Python may introduce additional formats.

.. _export-complexity:

Export complexity
-----------------

> In general, an export has a complexity of O(1): no memory copy is
> needed. There are cases when a copy is needed, O(n) complexity:

Reviewer: ```suggestion
In general, an export has a complexity of *O*\ (1): no memory copy is
needed. There are cases when a copy is needed, *O*\ (*n*) complexity:
```

> * If UCS2 is requested and the native format is UCS1.
> * If UCS4 is requested and the native format is UCS1 or UCS2.
> * If UTF8 is requested: the string is encoded to UTF-8 at the first

Reviewer: ```suggestion
* If UCS-2 is requested and the native format is UCS.
* If UCS-4 is requested and the native format is UCS or UCS-2.
* If UTF-8 is requested: the string is encoded to UTF-8 at the first
```

  call, and then the encoded UTF-8 string is cached.

> To have a O(1) complexity on CPython and PyPy, it's recommended to

Reviewer: ```suggestion
To have an *O*\ (1) complexity on CPython and PyPy, it's recommended to
```

``PyUnicode_FORMAT_ASCII``  ``"B"``             1 byte
==========================  ==================  ============

PyUnicode_Import()
------------------

API: ``PyObject* PyUnicode_Import(const void *data, Py_ssize_t nbytes, int32_t format)``.

Create a Unicode string object from a buffer in a supported format.

* Return a reference to a new string object on success.
* Set an exception and return ``NULL`` on error.

*data* must not be NULL. *nbytes* must be positive or zero.

See ``PyUnicode_Export()`` for the available formats.

Note: The ``PyUnicode_Import()`` function is similar to
``PyUnicode_FromKindAndData()``, but ``PyUnicode_FromKindAndData()`` is
excluded from the limited C API.

UTF-8 format
------------

CPython 3.14 doesn't use the UTF-8 format internally. The format is
provided for compatibility with PyPy which uses UTF-8 natively for
strings. Moreover, in CPython, the encoded UTF-8 string is cached which
makes it convenient to be exported.

On CPython, the UTF-8 format has the lowest priority: ASCII and UCS
formats are preferred.

ASCII format
------------

When the ``PyUnicode_FORMAT_ASCII`` format is request for export, the

> ``PyUnicode_FORMAT_UCS1`` export format is used for ASCII and Latin1

Reviewer: ```suggestion
``PyUnicode_FORMAT_UCS1`` export format is used for ASCII and Latin-1
```

strings.

The ``PyUnicode_FORMAT_ASCII`` format is mostly useful for
``PyUnicode_Import()`` to validate that the string only contains ASCII
characters.

Surrogate characters and NUL character
---------------------------------------

Surrogate characters are allowed: they can be imported and exported. For
example, the UTF-8 format uses the ``surrogatepass`` error handler.

Embedded NUL characters are allowed: they can be imported and exported.

An exported string does not end with a trailing NUL character: the
``PyUnicode_Export()`` caller must use ``Py_buffer.len`` to get the
string length.

Implementation
==============

https://github.com/python/cpython/pull/123738

Backwards Compatibility
=======================

There is no impact on the backward compatibility, only new C API
functions are added.

Usage of PEP 393 C APIs
=======================

> A code search on PyPI top 7,500 projects (on March 2024) shows that

Reviewer: ```suggestion
A code search on PyPI top 7,500 projects (in March 2024) shows that
```

----------------------

21 projects call ``PyUnicode_2BYTE_DATA()`` and/or
``PyUnicode_4BYTE_DATA()``:

* **Cython** (3.0.9)
* **MarkupSafe** (2.1.5)
* Nuitka (2.1.2)
* PyICU (2.12)
* PyICU-binary (2.7.4)
* PyQt5_sip (12.13.0)
* PyQt6_sip (13.6.0)
* biopython (1.83)
* catboost (1.2.3)
* cement (3.0.10)
* cffi (1.16.0)
* duckdb (0.10.0)
* **mypy** (1.9.0)
* **numpy** (1.26.4)
* orjson (3.9.15)
* pemja (0.4.1)
* pyahocorasick (2.0.0)
* pyjson5 (1.6.6)
* pyobjc-core (10.2)
* sip (6.8.3)
* wxPython (4.2.1)

Rejected Ideas
==============

Reject embedded NUL character and trailing NUL character
--------------------------------------------------------

In C, it's convenient to have a trailing NUL character. For example,
the ``for (; *str != 0; str++)`` loop can be used to iterate on
characters and ``strlen()`` can be used to get a string length.

The problem is that a Python ``str`` object can embed NUL characters.
Example: ``"ab\0c"``. If a string contains an embedded NUL character,
code relying on the NUL character to find the string end truncate the
string. It can lead to bugs, or even security vulnerabilities.

Rejecting embedded NUL characters require to scan the string which has

> a O(n) complexity.

Reviewer: ```suggestion
an *O*\ (*n*) complexity.
```

Reject surrogate characters
---------------------------

Surrogate characters are characters in the Unicode range [U+D800;
U+DFFF].  They are disallowed by UTF codecs such as UTF-8. A Python
``str`` object can contain arbitrary lone surrogate characters. Example:
``"\uDC80"``.

> Rejecting surogate characters prevents exporting a string which contain
> such character. It can be surprising and annoying since the
> ``PyUnicode_Export()`` caller doesn't control the string content.

Reviewer: ```suggestion
Rejecting surrogate characters prevents exporting a string which contains
such a character. It can be surprising and annoying since the
``PyUnicode_Export()`` caller doesn't control the string contents.
```
